### ISSUE TITLE
Python 3 client problem

### ISSUE THREAD

**Issue by smulube** [twitchtv/twirp#130]
This issue relates to https://github.com/twitchtv/twirp/pull/97 I believe.

I've been attempting to generate a python client to interact with a Twirp service written in Go and the generated client is unable to make requests to the server when running under Python 3 (tested with 3.6.6).

As a result of #97, the generated client now includes a line inside the constructor like:

```python
self.__target = server_address.encode('ascii')
```

Now under Python 2 this encode function returns another (encoded) string, however under Python 3, this function returns a distinct `bytes` type. This then breaks the generated `__make_request` function where we do something like this:

```python
req = Request(
    url=self.__target + '/twirp' + full_method,
    data=body,
    ...
```

On the line where we try to concatenate `self.__target` which is now a bytes object with the `'/twirp'` and full_method string, we get: `TypeError: can't concat str to bytes`.

For my purposes which is a single client that I can manually patch and I know is running Python 3, I can do something like this:

```python
req = Request(
    url=self.target.decode('ascii') + '/twirp' + full_method
    data=body,
    ...
```

or I suppose convert the other strings to byte types so they all append together nicely, or even revert the line that added that `encode('ascii')` call when initializing the client, but this feels like a bug to me.

I'm neither a Python or any kind of unicode expert so I don't what the _correct_ solution would be in terms of making this work properly under both Python 2 and 3, however I'd be happy to submit a PR if anyone could offer any guidance.

Tagging @daroot as the author of the #97.

**Comment by daroot** [comment]
Definitely a related problem.  I'm not a unicode expert either, and haven't written a ton of Python 3 that cares about unicode string vs byte strings and whether the httplib is safe to use with one or the other.  

I'm not 100% sure there's a simple solution that reliably works for both python2 and python3 that also guarantees passing something into their respective httplibs in a way that doesn't infect the resulting request and break the binary body.  

It may be necessary to actually break down and write some code that cares about python versions via sys.version or has some other logic for passing either a unicode or bytes type object to Request depending on what that version wants.

**Comment by smulube** [comment]
Have been playing with this in the python shell, I think in the generated client constructor we could either do something like:

```python
if sys.version_info[0] > 2:
    self.__target = server_address
else:
    self.__target = server_address.encode('ascii')
```

which is an attempt to follow the [porting guidelines](https://docs.python.org/3/howto/pyporting.html) which state:

> ... treat Python 2 as the exceptional case instead of Python 3 and assume that future Python versions will be more compatible with Python 3 than Python 2:

However those guidelines also state this:

> The best solution, though, is to do no version detection at all and instead rely on feature detection. That avoids any potential issues of getting the version detection wrong and helps keep you future-compatible:

or the alternative would be some feature detection way of dong this. I had a think about this, and I got to the following ugly thing:

```python
try:
    type(server_address) is unicode # this line does nothing but trigger a NameError in python 3
    self.__target = server_address.encode('ascii')
except NameError:
    self.__target = server_address
```

This is I think a "feature detection"  way of doing this, but is perhaps the reverse of what we want in that the feature I'm detecting here is the existence of the separate `unicode` type which I think only exists in Python 2 rather than detecting some Python 3 feature. I also really don't like the addition of that line of no purpose other than triggering an error in the generated code.

Happy to submit a PR with either of the above changes, but would love some feedback on which of the two approaches looks better, or perhaps a better suggestion?

**Comment by spenczar** [comment]
This should be fixed in #131. Thanks for working it out!

**Comment by smulube** [comment]
:+1:
